lib/metasploit/framework/password_crackers/cracker.rb

```diff
@@ -0,0 +1,452 @@
+          when 'mssql'
+            return '131'
+          when 'mssql05'
+            return '132'
+          when 'mssql12'
+            return '1731'
+          # hashcat requires a format we dont have all the data for
+          # in the current dumper, so this is disabled in module and lib
+          #when 'oracle', 'des,oracle'
+          #  return '3100'
+          when 'oracle11', 'raw-sha1,oracle'
+            return '112'
+          when 'oracle12c', 'pbkdf2,oracle12c'
+            return '12300'
+          when 'postgres', 'dynamic_1034', 'raw-md5,postgres'
+            return '12'
+          when 'mysql'
+            return '200'
+          when 'mysql-sha1'
+            return '300'
+          when 'PBKDF2-HMAC-SHA512' #osx 10.8+
+            return '7100'
+          when 'xsha' #osx 10.4-6
+            return '122'
+          when 'xsha512' #osx 10.7
+            return '1722'
+          when 'PBKDF2-HMAC-SHA1' #Atlassian
+            return '12001'
+          when 'phpass' #Wordpress/PHPass, Joomla, phpBB3
+            return '400'
+          when 'mediawiki' # mediawiki b type
+            return '3711'
+          end
+          nil
+        end
+
+
+        # This method sets the appropriate parameters to run a cracker in incremental mode
+        def mode_incremental
+          if cracker == 'john'
+            self.wordlist = nil
+            self.rules = nil
+            self.incremental = 'Digits'
+          elsif cracker == 'hashcat'
+            self.wordlist = nil
+            self.attack = '3'
+            self.incremental = true
+          end
+        end
+
+
+        # This method sets the appropriate parameters to run a cracker in wordlist mode
+        #
+        # @param[String] a file location of the wordlist to use
+        def mode_wordlist(file)
+          if cracker == 'john'
+            self.wordlist = file
+            self.rules = 'wordlist'
+            self.incremental = nil
+          elsif cracker == 'hashcat'
+            self.wordlist = file
+            self.attack = '0'
+            self.incremental = nil
+          end
+        end
+
+
+        # This method sets the john to 'normal' mode
+        def mode_normal
+          if cracker == 'john'
+            self.wordlist = nil
+            self.rules = nil
+            self.incremental = nil
+          end
+        end
+
+
+        # This method sets the john to single mode
+        #
+        # @param[String] a file location of the wordlist to use
+        def mode_single(file)
+          if cracker == 'john'
+            self.wordlist = file
+            self.rules = 'single'
+            self.incremental = nil
+          end
+        end
+
+
+        # This method follows a decision tree to determine the path
+        # to the cracker binary we should use.
+        #
+        # @return [NilClass] if a binary path could not be found
+        # @return [String] the path to the selected JtR binary
+        def binary_path
```

Comment:
Return early, return often, rather than spaghetti branches and EOL conditional:

```ruby
        def binary_path
          # Always prefer a manually entered path
          if cracker_path && ::File.file?(cracker_path)
            return cracker_path
          end

          # Look in the Environment PATH for the john binary
          if cracker == 'john'
            path = Rex::FileUtils.find_full_path("john") ||
              Rex::FileUtils.find_full_path("john.exe")
          elsif cracker == 'hashcat'
            path = Rex::FileUtils.find_full_path("hashcat") ||
              Rex::FileUtils.find_full_path("hashcat.exe")
          else
            raise PasswordCrackerNotFoundError, 'No suitable Cracker was selected, so a binary could not be found on the system'
          end

          if path && ::File.file?(path)
            return path
          end

          raise PasswordCrackerNotFoundError, 'No suitable john/hashcat binary was found on the system'
        end
```